fe/src/main/java/org/apache/doris/load/DeleteHandler.java

```diff
@@ -0,0 +1,549 @@
+        }
+
+        if (ok) {
+            commitJob(deleteJob, db, timeoutMs);
+        } else {
+            deleteJob.checkQuorum();
+            if (deleteJob.getState() != DeleteState.UN_QUORUM) {
+                long nowQuorumTimeMs = System.currentTimeMillis();
+                long endQuorumTimeoutMs = nowQuorumTimeMs + timeoutMs / 2;
+                // if job's state is finished or stay in quorum_finished for long time, try to commit it.
+                try {
+                    while (deleteJob.getState() == DeleteState.QUORUM_FINISHED && endQuorumTimeoutMs > nowQuorumTimeMs) {
+                        deleteJob.checkQuorum();
+                        Thread.sleep(1000);
+                        nowQuorumTimeMs = System.currentTimeMillis();
+                    }
+                    commitJob(deleteJob, db, timeoutMs);
+                } catch (InterruptedException e) {
+                }
+            } else {
+                List<Entry<Long, Long>> unfinishedMarks = countDownLatch.getLeftMarks();
+                // only show at most 5 results
+                List<Entry<Long, Long>> subList = unfinishedMarks.subList(0, Math.min(unfinishedMarks.size(), 5));
+                String errMsg = "Unfinished replicas:" + Joiner.on(", ").join(subList);
+                LOG.warn("delete job timeout: {}, {}", transactionId, errMsg);
+                cancelJob(deleteJob, "delete job timeout");
+                throw new DdlException("failed to delete replicas from job: {}, {}, transactionId, errMsg");
+            }
+        }
+    }
+
+    private void commitJob(DeleteJob job, Database db, long timeout) {
+        long transactionId = job.getTransactionId();
+        GlobalTransactionMgr globalTransactionMgr = Catalog.getCurrentGlobalTransactionMgr();
+        TransactionState transactionState = globalTransactionMgr.getTransactionState(transactionId);
+        List<TabletCommitInfo> tabletCommitInfos = new ArrayList<TabletCommitInfo>();
+        try {
+            TabletInvertedIndex invertedIndex = Catalog.getCurrentInvertedIndex();
+            for (TabletDeleteInfo tDeleteInfo : job.getTabletDeleteInfo()) {
+                for (Replica replica : tDeleteInfo.getFinishedReplicas()) {
+                    // the inverted index contains rolling up replica
+                    Long tabletId = invertedIndex.getTabletIdByReplica(replica.getId());
+                    if (tabletId == null) {
+                        LOG.warn("could not find tablet id for replica {}, the tablet maybe dropped", replica);
+                        continue;
+                    }
+                    tabletCommitInfos.add(new TabletCommitInfo(tabletId, replica.getBackendId()));
+                }
+            }
+            boolean isSuccess = globalTransactionMgr.commitAndPublishTransaction(db, transactionId, tabletCommitInfos, timeout);
+            if (!isSuccess) {
+                cancelJob(job, "delete timeout when waiting transaction commit");
```

Comment:
`!isSuccess` does not mean the transaction is failed, it means the transaction is committed and publish timeout (not VISIBLE yet). So you should not cancel the job, but should tell the user the delete is success but will be taking effect later.

Reply:
Done